### PR TITLE
This change adds ondemand prices

### PR DESCRIPTION
**Pull request by corlettb**
Most of the time when using spot prices you want to compare them to the ondemand cost for alerting and graphing.

Unfortunately aws does not provide an api for getting ondemand costs. In order for this to work I am pulling from https://raw.githubusercontent.com/powdahound/ec2instances.info/master/www/instances.json. This is the backend data from ec2instances.info. As this might not be 100% reliable I have only enabled this via a command line option which is off by default.

It is only updated once every 24 hours.

This change also adds logging and a verbose option.

I haven't deployed this to our environment yet. Only tested it locally. I thought I'd get your thoughts on this early.

### CODE REVIEW

**Comment by crashdump** [comment]
Thanks @corlettb, that would very helpful for me.

File: spot-price-monitor.py [hunk 1]
```diff
@@ -71,6 +106,10 @@ def get_args():
     parser.add_argument('-r', '--region', type=str, default='us-east-1',
                         help='''The region that the cluster is running
                         in (Default: us-east-1)''')
+    parser.add_argument('-o', '--ondemand', action="store_true", default=False,
```
Comment (dobesv) [inline spot-price-monitor.py]: Somehow I feel like `-o` should be for some kind of output since that what it's used for in kubernetes everywhere else.  I think it might be OK to have the long-form option only?

File: spot-price-monitor.py [hunk 2]
```diff
@@ -71,6 +106,10 @@ def get_args():
...
+                        help='''Will enable ondemand prices''')
+    parser.add_argument('-v', '--verbose', action="store_true", default=False,
+                        help='''Enable verbose output''')
```
Comment (dobesv) [inline spot-price-monitor.py]: I wonder if the logging changes are a bit out of scope here.  I'm not sure what the project owners wanted to do about logging, but maybe the debug text should just be removed from the PR and the errors can just be `print`ed ?

Not sure ... perhaps someone from pusher will chime in on this one.

**Comment by dobesv** [comment]
I've been thinking this would be a nice feature.  But it could also be published as a separate service.  In fact I bet you could create an AWS lambda job that just converts that file into prometheus format and we could put that behind cloudfront and point prometheus straight at it.

File: spot-price-monitor.py (hunk 1 above)
Comment (corlettb) [reply spot-price-monitor.py]: I'll happily remove the -o if thats what it takes to get it accepted

File: spot-price-monitor.py (hunk 2 above)
Comment (corlettb) [reply spot-price-monitor.py]: I certainly take the point that it might have been better in a different pull request. It is very useful though as it will also show debug messages from the kubernetes and aws api's.

**Comment by corlettb** [comment]
It could be included in a different service but to understand how "good" a spot price is you need to compare it to the ondemand price. Its just a shame you can't pull this info from the aws api.

**Comment by JoelSpeed** [comment]
@dobesv I initially thought that this could run as a separate service but ultimately the logic for filtering based on what's running in your cluster would be duplicated, so I think it makes sense to add this feature to the `spot-price-monitor`

@corlettb I've had a look over this and in general am happy with it. I'm gonna build an image and give it a test and I'll let you know how that goes.
Looks as though it shouldn't crash if the ec2 on demand info disappears so I think the external dependency (while annoying) is acceptable

**Comment by JoelSpeed** [comment]
@corlettb Could you please rebase this onto the latest master?

**Comment by corlettb** [comment]
I've rebased to the main line. I've also added a change to deal with the situation where '-l' is not used.

File: spot-price-monitor.py (hunk 1 above)
Comment (corlettb) [reply spot-price-monitor.py]: I've removed the -o short-form as discussed.

File: spot-price-monitor.py [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
 import boto3
+import requests
+import re
+import json
 from botocore.exceptions import ClientError
 from prometheus_client import start_http_server, Gauge, Counter
 from datetime import datetime
-from time import sleep
+import time
+import logging
```
Comment (dobesv) [inline spot-price-monitor.py]: If you are using the logging framework I wonder if the standard isn't

```
import logging
log = logging.getLogger(__name__)
```

File: spot-price-monitor.py [hunk 4]
```diff
@@ -112,26 +165,51 @@ def update_spot_price_metrics(metric, prices):
...
             backoff_multiplier = 1
         except ClientError as e:
-            error.label(code=e.response['Error']['Code']).inc()
+            error.labels(code=e.response['Error']['Code']).inc()
             if e.response['Error']['Code'] == 'RequestLimitExceeded':
                 backoff_multiplier *= 2
         update_spot_price_metrics(s, spot_prices)
-        sleep(args.scrape_interval * backoff_multiplier)
+
+        # refresh ondemand prices each hour
```
Comment (dobesv) [inline spot-price-monitor.py]: This comment says "each hour" but it looks like the time interval is actually one day

Comment (corlettb) [reply spot-price-monitor.py]: Fixed

File: spot-price-monitor.py [hunk 5]
```diff
@@ -80,6 +115,10 @@ def get_args():
     parser.add_argument('-r', '--region', type=str, default='us-east-1',
                         help='''The region that the cluster is running
                         in (Default: us-east-1)''')
+    parser.add_argument('--ondemand', action="store_true", default=False,
```
Comment (JoelSpeed) [inline spot-price-monitor.py]: Could you please change the flag name to be consistent with other multi-word flags?
```suggestion
    parser.add_argument('--on-demand', action="store_true", default=False,
```

File: spot-price-monitor.py (hunk 3 above)
Comment (JoelSpeed) [reply spot-price-monitor.py]: I agree with @dobesv but since this is internal it can be changed at a later date
